64_MergeSort/main.cpp

```diff
@@ -0,0 +1,368 @@
+				};
+				m_device->updateDescriptorSets(writes, {});
+			}
+
+			if (!m_outputBufferAllocation.memory->map({ 0ull,m_outputBufferAllocation.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
+			{
+				return logFail("Failed to map the Device Memory!\n");
+			}
+
+			if (!m_inputBufferAllocation.memory->map({ 0ull,m_inputBufferAllocation.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
+			{
+				return logFail("Failed to map the Device Memory!\n");
+			}
+		}
+
+		// Copy the random generated array to input buffer.
+		const int32_t* inputBufferMappedPointer = reinterpret_cast<const int32_t*>(m_inputBufferAllocation.memory->getMappedPointer());
+		memcpy((void*)inputBufferMappedPointer, inputBufferData.data(), sizeof(int32_t) * inputBufferData.size());
+
+		// Start of the merge sort algorithm.
+		const size_t numberOfPhases = (size_t)ceil(log2(NumberOfElementsToSort));
+
+		// A monotonically increasing counter value (for use by timeline semaphore).
+		uint64_t monotonicallyIncreasingCounter = 0u;
+		// Create the Semaphore (required since to start execution of phase X, phase X - 1 must complete execution on the GPU.
+		m_phaseSemaphore = m_device->createSemaphore(monotonicallyIncreasingCounter);
+
+		for (size_t phaseIndex = 1; phaseIndex <= numberOfPhases; phaseIndex++)
+		{
+			smart_refctd_ptr<nbl::video::IGPUCommandBuffer> cmdbuf;
+			{
+				smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(m_computeQueueFamily, IGPUCommandPool::CREATE_FLAGS::TRANSIENT_BIT);
+				if (!cmdpool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 1u, &cmdbuf))
+				{
+					return logFail("Failed to create Command Buffers!\n");
+				}
+			}
+
+			cmdbuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+			cmdbuf->beginDebugMarker("My Compute Dispatch", core::vectorSIMDf(0, 1, 0, 1));
+			cmdbuf->bindComputePipeline(m_computePipeline.get());
+
+			// If phase index is ODD, the inpt and output buffer are as usual. 
+			// If phase index is EVEN, input buffer is the output buffer of last phase.
+			if (phaseIndex % 2 == 1)
+			{
+				cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_pipelineLayout.get(), 0, 1, &m_descriptorSet1.get());
+			}
+			else
+			{
+				cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_pipelineLayout.get(), 0, 1, &m_descriptorSet2.get());
+			}
+
+			struct SortingPhaseData
+			{
+				uint32_t numElementsPerArray;
+				uint32_t bufferLength;
+			};
+
+			const SortingPhaseData pushConstantData = {
+				.numElementsPerArray = (uint32_t)pow(2, phaseIndex - 1),
+				.bufferLength = static_cast<uint32_t>(NumberOfElementsToSort),
+			};
+
+			cmdbuf->pushConstants(m_pipelineLayout.get(), nbl::asset::IShader::ESS_COMPUTE, 0u, sizeof(pushConstantData), &pushConstantData);
+
+			cmdbuf->dispatch((uint32_t)ceilf(NumberOfElementsToSort / powf(2.0f, phaseIndex)), 1, 1);
```

Comment:
why areyou using floats to do integer math !?

Reply:
Done, for getting integer ceil division of a / b I'm now using (a + b - 1) / b instead.